pkg/client/cache/controller.go

```diff
@@ -61,21 +61,23 @@ type Config struct {
 type ProcessFunc func(obj interface{}) error
 
 // Controller is a generic controller framework.
-type Controller struct {
+type controller struct {
 	config         Config
 	reflector      *Reflector
 	reflectorMutex sync.RWMutex
 }
 
 // TODO make the "Controller" private, and convert all references to use ControllerInterface instead
-type ControllerInterface interface {
+type Controller interface {
 	Run(stopCh <-chan struct{})
 	HasSynced() bool
+	LastSyncResourceVersion() string
+	Requeue(obj interface{}) error
```

Comment:
> Is this the controller with the "here there be dragons, trust us, its good" comment?

PV controller is in pretty good shape, it has >80% unit test coverage, the code is IMO well readable (I am biased, I wrote it) and I do not remember any big bug there since 1.3. That said, it's pretty complex piece of code (it solves complex problems!) and it has quite steep learning curve for anyone who wants to understand it.